nifi-extension-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/main/java/org/apache/nifi/services/protobuf/ProtobufSchemaCompiler.java

```diff
@@ -0,0 +1,245 @@
+import org.apache.nifi.serialization.record.SchemaIdentifier;
+
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import static java.nio.file.StandardOpenOption.CREATE;
+import static java.nio.file.StandardOpenOption.TRUNCATE_EXISTING;
+import static java.nio.file.StandardOpenOption.WRITE;
+import static org.apache.nifi.services.protobuf.ProtobufSchemaValidator.validateSchemaDefinitionIdentifiers;
+
+/**
+ * Handles Protocol Buffer schema compilation, caching, and temporary directory operations.
+ * This class is responsible for compiling schema definitions into Wire Schema objects,
+ * managing a cache of compiled schemas, and handling temporary directory operations
+ * required during the compilation process.
+ */
+final class ProtobufSchemaCompiler {
+
+    private static final List<Location> STANDARD_PROTOBUF_LOCATIONS = Arrays.asList(
+        Location.get(CoreLoaderKt.WIRE_RUNTIME_JAR, "google/protobuf/any.proto"),
+        Location.get(CoreLoaderKt.WIRE_RUNTIME_JAR, "google/protobuf/duration.proto"),
+        Location.get(CoreLoaderKt.WIRE_RUNTIME_JAR, "google/protobuf/empty.proto"),
+        Location.get(CoreLoaderKt.WIRE_RUNTIME_JAR, "google/protobuf/struct.proto"),
+        Location.get(CoreLoaderKt.WIRE_RUNTIME_JAR, "google/protobuf/timestamp.proto"),
+        Location.get(CoreLoaderKt.WIRE_RUNTIME_JAR, "google/protobuf/wrappers.proto")
+    );
+    private static final int CACHE_EXPIRE_HOURS = 1;
+    private static final int COMPILED_SCHEMAS_CACHE_SIZE = 200;
+
+    private final Cache<SchemaIdentifier, Schema> compiledSchemaCache;
+    private final ComponentLog logger;
+
+    /**
+     * Creates a new ProtobufSchemaCompiler with default cache settings.
+     *
+     * @param logger the component logger for logging compilation activities
+     */
+    public ProtobufSchemaCompiler(final ComponentLog logger) {
+        this.logger = logger;
+        this.compiledSchemaCache = Caffeine.newBuilder()
+            .expireAfterAccess(CACHE_EXPIRE_HOURS, TimeUnit.HOURS)
+            .maximumSize(COMPILED_SCHEMAS_CACHE_SIZE)
+            .build();
+    }
+
+    /**
+     * Compiles a schema definition or retrieves it from cache.
+     *
+     * @param schemaDefinition the schema definition to compile
+     * @return the compiled Schema
+     */
+    public Schema compileOrGetFromCache(final SchemaDefinition schemaDefinition) {
+        return compiledSchemaCache.get(schemaDefinition.getIdentifier(),
+            identifier -> {
+                try {
+                    return compileSchemaDefinition(schemaDefinition);
+                } catch (final IOException e) {
+                    throw new RuntimeException("Could not compile schema for identifier: " + identifier, e);
```

Comment:
`UncheckedIOException` should be preferred over the general `RuntimeException` when handling an `IOException`
```suggestion
                    throw new UncheckedIOException("Could not compile schema for identifier: " + identifier, e);
```

Reply:
Done